### PR TITLE
csi: change rook-ceph-csi-config to expose clusterID for subvolume

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

The rook-ceph-csi-config config map now includes CephFS subvolumegroups
when a subvolumegroup CR is created. It is useful for ceph-csi to
understand which subvolumegroup to use for a given cluster.

Signed-off-by: Sébastien Han <seb@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/file/subvolumegroup/controller.go [hunk 1]
```diff
@@ -272,9 +287,14 @@ func (r *ReconcileCephFilesystemSubVolumeGroup) updateStatus(client client.Clien
...
+	cephFilesystemSubVolumeGroup.Status.Info = map[string]string{"clusterID": buildClusterID(cephFilesystemSubVolumeGroup)}
 	if err := reporting.UpdateStatus(client, cephFilesystemSubVolumeGroup); err != nil {
 		logger.Errorf("failed to set ceph ceph filesystem subvolume group %q status to %q. %v", name, status, err)
 		return
 	}
 	logger.Debugf("ceph ceph filesystem subvolume group %q status updated to %q", name, status)
 }
+
+func buildClusterID(cephFilesystemSubVolumeGroup *cephv1.CephFilesystemSubVolumeGroup) string {
+	return fmt.Sprintf("%s-%s-file-%s", cephFilesystemSubVolumeGroup.Namespace, cephFilesystemSubVolumeGroup.Spec.FilesystemName, cephFilesystemSubVolumeGroup.Name)
```
Comment (Madhu-1) [inline pkg/operator/ceph/file/subvolumegroup/controller.go:299]: as discussed this should be hash or else we will hit clusterID length limit issue in cephcsi.

Comment (leseb) [reply pkg/operator/ceph/file/subvolumegroup/controller.go:299]: Or right, the hash 🤦🏻 , I can easily add this!

Comment (leseb) [reply pkg/operator/ceph/file/subvolumegroup/controller.go:299]: What's the char length limit already?

Comment (Madhu-1) [reply pkg/operator/ceph/file/subvolumegroup/controller.go]: clusterID length limit is 64 .  volID limit is 128 https://github.com/ceph/ceph-csi/blob/devel/internal/util/volid.go#L53

**Comment by Madhu-1** [comment]
@leseb dependency needs to be updated https://github.com/rook/rook/blob/master/design/ceph/resource-dependencies.md? if there are any PV/volumesnapshots in the subvolumegroup we should block/allow it for deletion?

File: pkg/operator/ceph/file/subvolumegroup/controller.go (hunk 1 above)
Comment (leseb) [reply pkg/operator/ceph/file/subvolumegroup/controller.go]: So 64 + 128 ?

Comment (Madhu-1) [reply pkg/operator/ceph/file/subvolumegroup/controller.go]: the clusterID max length is 64 means buildClusterID returned string should be less than 64

Comment (Madhu-1) [reply pkg/operator/ceph/file/subvolumegroup/controller.go]: why not always hash it instead of hashing only if the length is exceeding?

Comment (leseb) [reply pkg/operator/ceph/file/subvolumegroup/controller.go]: I thought we said it would be useful to see a non-hashed value if possible.

Comment (Madhu-1) [reply pkg/operator/ceph/file/subvolumegroup/controller.go]: yeah, I remember but having hash always avoids confusion (the clusterID always remains hash in whatever the case) even if a user creates the filesystem/subvolumegroup/cluster namespace with bigger or smaller names?

**Comment by leseb** [comment]
> @leseb dependency needs to be updated https://github.com/rook/rook/blob/master/design/ceph/resource-dependencies.md? if there are any PV/volumesnapshots in the subvolumegroup we should block/allow it for deletion?

Right now, the deletion is blocked already if Ceph detects that there are subvolumes in the subvolumegroup, so it's sufficient.

**Comment by Madhu-1** [comment]
> > @leseb dependency needs to be updated https://github.com/rook/rook/blob/master/design/ceph/resource-dependencies.md? if there are any PV/volumesnapshots in the subvolumegroup we should block/allow it for deletion?
> 
> Right now, the deletion is blocked already if Ceph detects that there are subvolumes in the subvolumegroup, so it's sufficient.

@leseb  I don't see a code for this one if you don't mind can you please point me to that one?

**Comment by leseb** [comment]
> > > @leseb dependency needs to be updated https://github.com/rook/rook/blob/master/design/ceph/resource-dependencies.md? if there are any PV/volumesnapshots in the subvolumegroup we should block/allow it for deletion?
> > 
> > 
> > Right now, the deletion is blocked already if Ceph detects that there are subvolumes in the subvolumegroup, so it's sufficient.
> 
> @leseb I don't see a code for this one if you don't mind can you please point me to that one?

https://github.com/rook/rook/blob/376ca62f8ad07540d9ddffe9dc0ee53f4ac35e29/pkg/operator/ceph/file/subvolumegroup/controller.go#L244-L253

Also, I just realized there is a small issue with this code which I'll correct in this PR.

File: pkg/operator/ceph/file/subvolumegroup/controller.go (hunk 1 above)
Comment (leseb) [reply pkg/operator/ceph/file/subvolumegroup/controller.go]: Changed to use a hash only

**Review by Madhu-1 — APPROVED** [review]
LGTM

File: pkg/operator/ceph/csi/cluster_config.go [hunk 2]
```diff
@@ -155,24 +174,24 @@ func CreateCsiConfigMap(namespace string, clientset kubernetes.Interface, ownerI
 // value that is provided to ceph-csi uses in the storage class.
 // The locker l is typically a mutex and is used to prevent the config
 // map from being updated for multiple clusters simultaneously.
-func SaveClusterConfig(
-	clientset kubernetes.Interface, clusterNamespace string,
-	clusterInfo *cephclient.ClusterInfo, l sync.Locker) error {
+func SaveClusterConfig(clientset kubernetes.Interface, clusterNamespace string, clusterInfo *cephclient.ClusterInfo, l sync.Locker, newCsiClusterConfigEntry CsiClusterConfigEntry) error {
```
Comment (travisn) [inline pkg/operator/ceph/csi/cluster_config.go:177]: How about we use a package var to lock? If callers don't use the same lock, different controllers could attempt to update the configmap at the same time, right?
```suggestion
func SaveClusterConfig(clientset kubernetes.Interface, clusterNamespace string, clusterInfo *cephclient.ClusterInfo, newCsiClusterConfigEntry CsiClusterConfigEntry) error {
```

File: pkg/operator/ceph/csi/cluster_config.go [hunk 3]
```diff
@@ -89,32 +95,45 @@ func monEndpoints(mons map[string]*cephclient.MonInfo) []string {
...
 	cc, err := parseCsiClusterConfig(curr)
 	if err != nil {
 		return "", errors.Wrap(err, "failed to parse current csi cluster config")
 	}
 
 	for i, centry := range cc {
 		if centry.ClusterID == clusterKey {
-			centry.Monitors = monEndpoints(mons)
+			centry.Monitors = newCsiClusterConfigEntry.Monitors
+			if newCsiClusterConfigEntry.CephFS != nil && newCsiClusterConfigEntry.CephFS.SubvolumeGroup != "" {
```
Comment (travisn) [inline pkg/operator/ceph/csi/cluster_config.go:113]: If the subvolume group was set by the filesystem controller, but not set by the main cluster controller (when reconciling the mons), the subvolume group will still be preserved here, right? How about a comment about that?

File: pkg/operator/ceph/csi/cluster_config_test.go [hunk 4]
```diff
@@ -19,71 +19,98 @@ package csi
...
-	assert.Equal(t, len(cc), 2)
-	assert.Equal(t, cc[0].ClusterID, "alpha")
-	assert.Contains(t, cc[0].Monitors, "1.2.3.4:5000")
-	assert.Contains(t, cc[0].Monitors, "10.11.12.13:5000")
-	assert.Equal(t, len(cc[0].Monitors), 2)
-	assert.Equal(t, cc[1].ClusterID, "beta")
-	assert.Contains(t, cc[1].Monitors, "20.1.1.1:5000")
-	assert.Contains(t, cc[1].Monitors, "20.1.1.2:5000")
-	assert.Equal(t, len(cc[1].Monitors), 2)
+	t.Run("add a 3rd cluster with subvolumegroup", func(t *testing.T) {
```
Comment (travisn) [inline pkg/operator/ceph/csi/cluster_config_test.go:92]: Is there a unit test that shows the subvolume groups are preserved if the mons are updated?

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/cluster_config_test.go (hunk 4 above)
Comment (leseb) [reply pkg/operator/ceph/csi/cluster_config_test.go:92]: Added

File: pkg/operator/ceph/csi/cluster_config.go (hunk 3 above)
Comment (leseb) [reply pkg/operator/ceph/csi/cluster_config.go:113]: Added

File: pkg/operator/ceph/csi/cluster_config.go (hunk 2 above)
Comment (travisn) [reply pkg/operator/ceph/csi/cluster_config.go:182]: I see that callers are passing `&csi.ConfigMutex`. Why not make the mutex a private package var and no need to pass it as a param? This method could just use the package var directly, right?

**Comment by leseb** [comment]
@Madhu-1 if we have the hash for everything, we will have leftover entries for the same cluster, see:

```
  csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["10.107.84.70:6789","10.97.224.243:6789","10.111.155.211:6789"]},{"clusterID":"4df0bb6e1b7ddc0c1ea080cbe043dd1a","monitors":["10.97.224.243:6789","10.111.155.211:6789","10.107.84.70:6789"],"cephFS":{"subvolumeGroup":"group-a"}}]'
```

For this, should we only apply the hash once a certain length is reached? Thanks!

**Comment by Madhu-1** [comment]
> @Madhu-1 if we have the hash for everything, we will have leftover entries for the same cluster, see:
> 
> ```
>   csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["10.107.84.70:6789","10.97.224.243:6789","10.111.155.211:6789"]},{"clusterID":"4df0bb6e1b7ddc0c1ea080cbe043dd1a","monitors":["10.97.224.243:6789","10.111.155.211:6789","10.107.84.70:6789"],"cephFS":{"subvolumeGroup":"group-a"}}]'
> ```
> 
> For this, should we only apply the hash once a certain length is reached? Thanks!

@leseb sorry i didn't get the problem with leftover. can you please explain it?

**Comment by leseb** [comment]
> > @Madhu-1 if we have the hash for everything, we will have leftover entries for the same cluster, see:
> > ```
> >   csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["10.107.84.70:6789","10.97.224.243:6789","10.111.155.211:6789"]},{"clusterID":"4df0bb6e1b7ddc0c1ea080cbe043dd1a","monitors":["10.97.224.243:6789","10.111.155.211:6789","10.107.84.70:6789"],"cephFS":{"subvolumeGroup":"group-a"}}]'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For this, should we only apply the hash once a certain length is reached? Thanks!
> 
> @leseb sorry i didn't get the problem with leftover. can you please explain it?

Essentially we have duplicate entries, the old once with clusterID: rook-ceph and the new clusterID with the hash as a value.

**Comment by Madhu-1** [comment]
> > > @Madhu-1 if we have the hash for everything, we will have leftover entries for the same cluster, see:
> > > ```
> > >   csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["10.107.84.70:6789","10.97.224.243:6789","10.111.155.211:6789"]},{"clusterID":"4df0bb6e1b7ddc0c1ea080cbe043dd1a","monitors":["10.97.224.243:6789","10.111.155.211:6789","10.107.84.70:6789"],"cephFS":{"subvolumeGroup":"group-a"}}]'
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > For this, should we only apply the hash once a certain length is reached? Thanks!
> > 
> > 
> > @leseb sorry i didn't get the problem with leftover. can you please explain it?
> 
> Essentially we have duplicate entries, the old one with clusterID: rook-ceph and the new clusterID with the hash as a value.

Sorry, I missed one thing in review. the configmap entry should be deleted when the subvolumegroup CR is deleted. is this already handled?

**Comment by leseb** [comment]
> > > > @Madhu-1 if we have the hash for everything, we will have leftover entries for the same cluster, see:
> > > > ```
> > > >   csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["10.107.84.70:6789","10.97.224.243:6789","10.111.155.211:6789"]},{"clusterID":"4df0bb6e1b7ddc0c1ea080cbe043dd1a","monitors":["10.97.224.243:6789","10.111.155.211:6789","10.107.84.70:6789"],"cephFS":{"subvolumeGroup":"group-a"}}]'
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > For this, should we only apply the hash once a certain length is reached? Thanks!
> > > 
> > > 
> > > @leseb sorry i didn't get the problem with leftover. can you please explain it?
> > 
> > 
> > Essentially we have duplicate entries, the old one with clusterID: rook-ceph and the new clusterID with the hash as a value.
> 
> Sorry, I missed one thing in review. the configmap entry should be deleted when the subvolumegroup CR is deleted. is this already handled?

No, let me fix this.

File: pkg/operator/ceph/csi/cluster_config.go [hunk 5]
```diff
@@ -33,15 +33,22 @@ import (
 )
 
 var (
-	logger = capnslog.NewPackageLogger("github.com/rook/rook", "ceph-csi")
+	logger      = capnslog.NewPackageLogger("github.com/rook/rook", "ceph-csi")
+	ConfigMutex sync.Mutex
```
Comment (travisn) [inline pkg/operator/ceph/csi/cluster_config.go:37]: Can this be renamed or is it still used outside the cluster?
```suggestion
	configMutex sync.Mutex
```

**Review by travisn — CHANGES_REQUESTED** [review]
A small suggestion, and the linter hit an issue.

File: pkg/operator/ceph/cluster/mon/mon.go [hunk 6]
```diff
@@ -175,8 +175,7 @@ func New(context *clusterd.Context, namespace string, spec cephv1.ClusterSpec, o
 		mapping: &Mapping{
 			Schedule: map[string]*MonScheduleInfo{},
 		},
-		ownerInfo:      ownerInfo,
-		csiConfigMutex: csiConfigMutex,
+		ownerInfo: ownerInfo,
```
Comment (Madhu-1) [inline pkg/operator/ceph/cluster/mon/mon.go:178]: structCheck is failing as `csiConfigMutex` is not used anywhere.

**Review by travisn — APPROVED** [review]

**Pull request merged**